### PR TITLE
Automate device exists scenario

### PR DESCRIPTION
**Pull request by vprabhakar-px**
automating device removal issue

### CODE REVIEW

File: tests/basic/customer_issues_test.go [hunk 1]
```diff
@@ -757,3 +757,185 @@ var _ = Describe("{CreateCloudSnapAndDelete}", func() {
 		AfterEachTest(contexts)
 	})
 })
+
+func getVolumeAttachedNodeAndKill(volName *volume.Volume) (string, *node.Node, error) {
```
Comment (lsrinivas-pure) [inline tests/basic/customer_issues_test.go:761]: can you just return node as it has node ip/name details also

File: tests/basic/customer_issues_test.go [hunk 2]
```diff
@@ -757,3 +757,185 @@ var _ = Describe("{CreateCloudSnapAndDelete}", func() {
...
+
+	killPxStorageErr := KillPxStorageUsingPid(nodeDetails)
+	if killPxStorageErr != nil {
+		return "", nil, err
+	}
+
+	return attachedNode, &nodeDetails, nil
+}
+
+func isPodStuckNotRunning(nameSpace string) (bool, map[string]string, error) {
```
Comment (lsrinivas-pure) [inline tests/basic/customer_issues_test.go:787]: you can use IsPxRunningOnNode function

File: tests/basic/customer_issues_test.go [hunk 3]
```diff
@@ -757,3 +757,185 @@ var _ = Describe("{CreateCloudSnapAndDelete}", func() {
...
+		log.Infof("Volume picked for testing device exist issue [%v]", randomVol)
+
+		appVol, err := Inst().V.InspectVolume(randomVol.ID)
+		allNodesForVolume := appVol.ReplicaSets[0].Nodes
+		log.InfoD("List of nodes for the volume [%s] are [%v]", randomVol.Name, allNodesForVolume)
+
+		go func(nameSpace string) {
+			defer GinkgoRecover()
+			for {
+				podRestating, restartNode, err := isPodStuckNotRunning(nameSpace)
```
Comment (lsrinivas-pure) [inline tests/basic/customer_issues_test.go:853]: nit: podRestarting

File: tests/basic/customer_issues_test.go [hunk 4]
```diff
@@ -757,3 +757,185 @@ var _ = Describe("{CreateCloudSnapAndDelete}", func() {
...
+		}
+		log.Infof("Volume picked for testing device exist issue [%v]", randomVol)
+
+		appVol, err := Inst().V.InspectVolume(randomVol.ID)
+		allNodesForVolume := appVol.ReplicaSets[0].Nodes
+		log.InfoD("List of nodes for the volume [%s] are [%v]", randomVol.Name, allNodesForVolume)
+
+		go func(nameSpace string) {
+			defer GinkgoRecover()
+			for {
```
Comment (lsrinivas-pure) [inline tests/basic/customer_issues_test.go:852]: seems there is no break in this loop. please have a look

File: tests/common.go [hunk 5]
```diff
@@ -10242,3 +10242,74 @@ func GetNodeForGivenVolumeName(volName string) (*node.Node, error) {
...
+	log.InfoD("killing PID using command [%s]", command)
+	err = runCmd(command, memberNode)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// KillPxStorageUsingPid return error in case of command failure
+func KillPxStorageUsingPid(memberNode node.Node) error {
```
Comment (lsrinivas-pure) [inline tests/common.go:10290]: u can use stopDriver function with force option

File: tests/basic/customer_issues_test.go (hunk 1 above)
Comment (vprabhakar-px) [reply tests/basic/customer_issues_test.go]: done

File: tests/basic/customer_issues_test.go (hunk 4 above)
Comment (vprabhakar-px) [reply tests/basic/customer_issues_test.go:852]: done

File: tests/common.go (hunk 5 above)
Comment (vprabhakar-px) [reply tests/common.go:10246]: done

File: tests/basic/customer_issues_test.go [hunk 6]
```diff
@@ -757,3 +757,195 @@ var _ = Describe("{CreateCloudSnapAndDelete}", func() {
...
+	})
+
+	var contexts []*scheduler.Context
+	stepLog := "App Stuck in ContainerCreation State with Device Exists in the backend"
+	It(stepLog, func() {
+
+		var isPodRestarting bool = false
+		var terminateAll bool = false
+
+		Inst().AppList = []string{"vdbench-sv4-svc"}
```
Comment (lsrinivas-pure) [inline tests/basic/customer_issues_test.go:822]: can you add more apps or else we can use the logic  I have  used in PR
https://github.com/portworx/torpedo/pull/2020/files

File: tests/basic/customer_issues_test.go [hunk 7]
```diff
@@ -757,3 +757,195 @@ var _ = Describe("{CreateCloudSnapAndDelete}", func() {
...
+		contexts = make([]*scheduler.Context, 0)
+		for i := 0; i < Inst().GlobalScaleFactor; i++ {
+			contexts = append(contexts, ScheduleApplications(fmt.Sprintf("volwithdeviceexist-%d", i))...)
+		}
+		ValidateApplications(contexts)
+		//defer appsValidateAndDestroy(contexts)
```
Comment (lsrinivas-pure) [inline tests/basic/customer_issues_test.go:828]: remove this if not needed

File: tests/basic/customer_issues_test.go [hunk 8]
```diff
@@ -757,3 +757,195 @@ var _ = Describe("{CreateCloudSnapAndDelete}", func() {
...
+
+		go func(nameSpace string) {
+			defer GinkgoRecover()
+			for {
+				podRestarting, restartNode, err := isPodStuckNotRunning(nameSpace)
+				if podRestarting && len(restartNode) > 1 {
+					isPodRestarting = true
+				}
+
+				if err != nil {
```
Comment (lsrinivas-pure) [inline tests/basic/customer_issues_test.go:861]: if check not required

File: tests/basic/customer_issues_test.go [hunk 9]
```diff
@@ -757,3 +757,195 @@ var _ = Describe("{CreateCloudSnapAndDelete}", func() {
...
+			if isPodRestarting == false {
+				rand.Seed(time.Now().UnixNano())
+				randomNumber := rand.Intn(3) + 1
+
+				nodeDetail, err := getVolumeAttachedNodeAndKill(randomVol)
+				log.FailOnError(err, "Failed to kill Px Daemons")
+				log.InfoD("Volume  is Attached on Node {%s}", nodeDetail.GetMgmtIp())
+				restartedNode = append(restartedNode, *nodeDetail)
+
+				if randomNumber == 3 {
```
Comment (lsrinivas-pure) [inline tests/basic/customer_issues_test.go:896]: can you add a comment here explaining the logic

**Review by lsrinivas-pure — APPROVED** [review]
LGTM

**Pull request merged**